casepro/statistics/models.py

```diff
@@ -0,0 +1,165 @@
+                """
+
+                params = [unsquashed[f] for f in unique_fields]
+                cursor.execute(sql, params + params)
+
+        max_id = cls.objects.order_by('-pk').values_list('pk', flat=True).first()
+        if max_id:
+            cache.set(last_squash_key, max_id)
+
+    @classmethod
+    def get_by_org(cls, orgs, item_type, since=None, until=None):
+        return cls._get_count_set(item_type, {cls.encode_scope(o): o for o in orgs}, since, until)
+
+    @classmethod
+    def get_by_partner(cls, partners, item_type, since=None, until=None):
+        return cls._get_count_set(item_type, {cls.encode_scope(p): p for p in partners}, since, until)
+
+    @classmethod
+    def get_by_user(cls, org, users, item_type, since=None, until=None):
+        return cls._get_count_set(item_type, {cls.encode_scope(org, u): u for u in users}, since, until)
+
+    @classmethod
+    def _get_count_set(cls, item_type, scopes, since, until):
+        counts = cls.objects.filter(item_type=item_type)
+        if scopes:
+            counts = counts.filter(scope__in=scopes.keys())
+        if since:
+            counts = counts.filter(day__gte=since)
+        if until:
+            counts = counts.filter(day__lt=until)
+        return DailyCount.CountSet(counts, scopes)
+
+    class CountSet(object):
+        """
+        A queryset of counts which can be aggregated in different ways
+        """
+        def __init__(self, counts, scopes):
+            self.counts = counts
+            self.scopes = scopes
+
+        def total(self):
+            """
+            Calculates the overall total over a set of counts
+            """
+            total = self.counts.aggregate(total=Sum('count'))
+            return total['total'] if total['total'] is not None else 0
+
+        def day_totals(self):
+            """
+            Calculates per-day totals over a set of counts
+            """
+            return list(self.counts.values_list('day').annotate(total=Sum('count')).order_by('day'))
+
+        def month_totals(self):
+            """
+            Calculates per-month totals over a set of counts
+            """
+            counts = self.counts.extra(select={'month': 'EXTRACT(month FROM "day")'})
+            return list(counts.values_list('month').annotate(replies=Sum('count')).order_by('month'))
+
+        def scope_totals(self):
+            """
+            Calculates per-scope totals over a set of counts
+            """
+            totals = list(self.counts.values_list('scope').annotate(replies=Sum('count')))
+            total_by_encoded_scope = {t[0]: t[1] for t in totals}
+
+            total_by_scope = {}
+            for encoded_scope, scope in six.iteritems(self.scopes):
+                total_by_scope[scope] = total_by_encoded_scope.get(encoded_scope, 0)
+
+            return total_by_scope
+
+    class Meta:
+        index_together = ('item_type', 'scope', 'day')
+
+
+def record_new_outgoing(outgoing):
```

Comment:
Idea is that this records an outgoing message for the whole statistics module which might do more than just make instances of `DailyCount`